File: 04-trees-and-graphs/03_bfs_dfs_graph/bfs_dfs_graph.cpp
/**
 * Problem: Graph Traversals (BFS & DFS)
 * 
 * Representation: Adjacency List
 * Time Complexity: O(V + E)
 * Space Complexity: O(V + E)
 */

#include <iostream>
#include <vector>
#include <queue>

class Graph {
    int V;
    std::vector<std::vector<int>> adj;

    void dfsUtil(int v, std::vector<bool>& visited) {
        visited[v] = true;
        std::cout << v << " ";
        for (int neighbor : adj[v]) {
            if (!visited[neighbor]) {
                dfsUtil(neighbor, visited);
            }
        }
    }

public:
    Graph(int v) : V(v), adj(v) {}

    void addEdge(int u, int v) {
        adj[u].push_back(v);
        adj[v].push_back(u); // undirected
    }

    void bfs(int start) {
        std::vector<bool> visited(V, false);
        std::queue<int> q;

        visited[start] = true;
        q.push(start);

        std::cout << "BFS: ";
        while (!q.empty()) {
            int u = q.front();
            q.pop();
            std::cout << u << " ";

            for (int v : adj[u]) {
                if (!visited[v]) {
                    visited[v] = true;
                    q.push(v);
                }
            }
        }
        std::cout << "\n";
    }

    void dfs(int start) {
        std::vector<bool> visited(V, false);
        std::cout << "DFS: ";
        dfsUtil(start, visited);
        std::cout << "\n";
    }
};

int main() {
    std::cout << "[C++] Graph Traversals Test\n";
    Graph g(5);
    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 3);
    g.addEdge(1, 4);

    g.bfs(0);
    g.dfs(0);
    return 0;
}

File: 04-trees-and-graphs/03_bfs_dfs_graph/bfs_dfs_graph.py
"""
Problem: Graph Traversals (BFS & DFS)

Representation: Adjacency List (Dict of Lists)
Time Complexity: O(V + E)
Space Complexity: O(V + E)
"""

from collections import deque
from typing import Dict, List

class Graph:
    def __init__(self):
        self.adj: Dict[int, List[int]] = {}

    def add_edge(self, u: int, v: int):
        self.adj.setdefault(u, []).append(v)
        self.adj.setdefault(v, []).append(u)

    def bfs(self, start: int) -> List[int]:
        visited = set([start])
        queue = deque([start])
        order = []

        while queue:
            node = queue.popleft()
            order.append(node)
            for neighbor in self.adj.get(node, []):
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
        return order

    def dfs(self, start: int) -> List[int]:
        visited = set()
        order = []

        def _dfs(node):
            visited.add(node)
            order.append(node)
            for neighbor in self.adj.get(node, []):
                if neighbor not in visited:
                    _dfs(neighbor)

        _dfs(start)
        return order

if __name__ == "__main__":
    print("[Python] Graph Traversals Test")
    g = Graph()
    g.add_edge(0, 1)
    g.add_edge(0, 2)
    g.add_edge(1, 3)
    g.add_edge(1, 4)

    bfs_res = g.bfs(0)
    dfs_res = g.dfs(0)
    print("BFS Order:", bfs_res)
    print("DFS Order:", dfs_res)
    assert len(bfs_res) == 5
    assert len(dfs_res) == 5
    print("Graph traversals test passed!")
